### ISSUE TITLE
GitHub "Android CI" is unreliable.

### ISSUE THREAD

**Issue by mnalis** [commons-app/apps-android-commons#5245]
### Summary

"Android CI" github workflow fails in majority of cases (11 out of 12 cases for me) with errors like `java.lang.IllegalStateException: init() must be called prior to using this method.` eg [run 9](https://github.com/mnalis/apps-android-commons/actions/runs/5324588065) or [run 10](https://github.com/mnalis/apps-android-commons/actions/runs/5324741996) or [run 11](https://github.com/mnalis/apps-android-commons/actions/runs/5325271365) until it finally succeeds (**without any code changes**!) in [run 12](https://github.com/mnalis/apps-android-commons/actions/runs/5325276682). 
(I've even added `[workflow_dispatch]` trigger so it can be triggered manually without _any_ repo changes)

- Note that is in latest master (based upon https://github.com/commons-app/apps-android-commons/commit/f4e8de79fba592ed8f5572878768bf19c7da658c) after fixes in https://github.com/commons-app/apps-android-commons/pull/5236, and https://github.com/commons-app/apps-android-commons/pull/5237 and that the repository content is the same in cases when it works, and in cases when it fails. 

- some earlier runs like [run 3](https://github.com/mnalis/apps-android-commons/actions/runs/5291266502) seem to fail with slightly different error `androidx.test.espresso.base.RootViewPicker$RootViewWithoutFocusException: Waited for the root of the view hierarchy to have window focus and not request layout for 10 seconds. If you specified a non default root matcher, it may be picking a root that never takes focus.` , but those did not have https://github.com/commons-app/apps-android-commons/pull/5241 included yet.

It seems to be a race condition triggered depending on the GitHub servers load or something. The solution would probably be trying to idle for some time while re-checking if the testing environment has become usable, before proceeding with the tests.

### Steps to reproduce

1. fork the repository
2. push some data to the fork
3. notice that "Android CI" github workflow is run, but fails majority of the times

### Expected behaviour

"Android CI" github workflow should succeed reliably and reproducibly (assuming there are no bugs in the code, of course!). In any case, its behaviour should be deterministic - for same codebase, it should either **always** succeed or **always** fail.

### Actual behaviour

"Android CI" github workflow fails with many failed test:

```
fr.free.nrw.commons.MainActivityTest > testExplore[test(AVD) - 7.0] FAILED 
	java.lang.IllegalStateException: init() must be called prior to using this method.
	at androidx.test.espresso.intent.Checks.checkState(Checks.java:70)
```

### Android version

n/a

### Commons app version

n/a

### Would you like to work on the issue?

None

**Comment by mnalis** [comment]
There is closed issue https://github.com/commons-app/apps-android-commons/issues/4958 that mentions similar problem from before; which seems to have attempted a workaround by not running tests in some cases.

**Comment by sivaraam** [comment]
> 2. Push some data to the fork

Could you elaborate about this one? Is there any chnge that might affect the CI run? 🤔

**Comment by mnalis** [comment]
>>   Push some data to the fork
> Could you elaborate about this one? Is there any chnge that might affect the CI run? :thinking:

No, there shouldn't be any change that would affect those failures. In detail:

- same problem arises even on initial fork ([run 1](https://github.com/mnalis/apps-android-commons/actions/runs/5291262919)) when everything is same (by definition)
- the reason why some push (or PR) is needed is because [android.yml](https://github.com/commons-app/apps-android-commons/blob/master/.github/workflows/android.yml) specifies that CI runs only on `on: [push, pull_request]`, so otherwise I couldn't force it to run multiple times to rule-out one-off failures
- So I've changed that line to say `on: [push, pull_request, workflow_dispatch]` so there is manual button in GitHub `Actions` tab, so I can retry it as many times as needed. (As it is generally useful, and not just for this test, I've made an one-liner PR for that: https://github.com/commons-app/apps-android-commons/pull/5248)
- then I've just after each failed build **clicked that button** to run new build again **without any repository changes**, until it finally succeeded after several retries.

**Comment by rohit9625** [comment]
The CI fails only on our forked repo not on the main repo. So, I don't think this is an issue at all.
What do you think Mr. @sivaraam?

**Comment by mnalis** [comment]
> The CI fails only on our forked repo not on the main repo. So, I don't think this is an issue at all.

Why wouldn't that be an issue?! There is only one original repo, and more then 1200 forked ones. Certainly that is all the more reasons to fix it, then if that was just one broken repo and 1200 working ones  :smiling_face_with_tear: 

It still fails for me: https://github.com/mnalis/apps-android-commons/actions/runs/7646438188/job/20835243644

Perhaps those `java.lang.IllegalStateException: init() must be called prior to using this method` are related to this:
https://stackoverflow.com/questions/33624802/android-espresso-intents-test-randomly-fail-with-init-must-be-called-prior-t

**Comment by rohit9625** [comment]
Well, I have faced this error before and just now. See [here](https://github.com/rohit9625/apps-android-commons/actions/runs/7650104472/job/20845652824)

I thought that it was happening because my repo was forked from the main and my workflow file was not configured to run on my repo. And if I am wrong then the issue needs to be fixed.

Sorry, I misunderstood the issue.

**Comment by sivaraam** [comment]
> I thought that it was happening because my repo was forked from the main and my workflow file was not configured to run on my repo. And if I am wrong then the issue needs to be fixed.

The workflow file should be generic of the repo, Rohit. The workflow file doesn't need any tweaks when a repo has been forked, ideally.

The CI is actually failing our repo itself [ [ref](https://github.com/commons-app/apps-android-commons/actions/runs/7681808057/job/20935371210) ]. We indeed need to check the same and fix it.

**Comment by rohit9625** [comment]
Yes, you are right. The problem is with the instrumentation tests. I am investigating why is this failling.

**Comment by nicolas-raoul** [comment]
Do coverage checks run successfully on aonyone's environment?

**Comment by rohit9625** [comment]
I want to run some tests but this code throwing the exception:-
`Configure your beta account's username`

Can you anyone help me in configuration of username : )

**Comment by sivaraam** [comment]
> Do coverage checks run successfully on aonyone's environment?

I suppose not. I think issue mentioned in #5324 still exists.

> Can anyone help me with the configuration of the username?

Could you let us know how you ran the test?

**Comment by RitikaPahwa4444** [comment]
Hi @rohit9625, please add these properties to your `local.properties` file (add appropriate values):  
- `TEST_USER_NAME=<your_beta_account_username>`
- `TEST_USER_PASSWORD=<your_password>`
Also, please avoid doing a `git add .` if local.properties shows up as one of the changes because passwords should not be committed. 

Hope this helps! :)

**Comment by rohit9625** [comment]
Thank you @RitikaPahwa4444 
For providing these details :)
